File: sol/2115/2115.py
from collections import defaultdict

class Solution:
    def findAllRecipes(self, recipes: List[str], ingredients: List[List[str]], supplies: List[str]) -> List[str]:
        availables = frozenset(supplies)
        n = len(recipes)
        indeg = [0] * n
        leadsTo = defaultdict(list)
        canMake = []
        for i in range(n):
            for ing in ingredients[i]:
                if ing in availables:
                    continue
                indeg[i] += 1
                leadsTo[ing].append(i)
            if indeg[i] == 0:
                canMake.append(i)
        products = []
        while len(canMake) != 0:
            target = recipes[canMake.pop()]
            products.append(target)
            for newIdx in leadsTo[target]:
                indeg[newIdx] -= 1
                if indeg[newIdx] == 0:
                    canMake.append(newIdx)
        return products

File: sol/2115/2115.cpp
#include <unordered_map>
#include <unordered_set>

struct Solution {
    vector<string> findAllRecipes(vector<string>& recipes, vector<vector<string>>& ingredients, vector<string>& supplies) {
        int n = recipes.size();
        unordered_set<string> suppSet(supplies.begin(), supplies.end());
        vector<int> indeg(n, 0), canMake;
        unordered_map<string, vector<int>> leadsTo;
        for (int i = 0; i < n; ++i){
            for (auto &ing : ingredients[i]){
                if (suppSet.contains(ing)){
                    continue;
                }
                ++indeg[i];
                leadsTo[ing].push_back(i);
            }
            if (indeg[i] == 0){
                canMake.push_back(i);
            }
        }
        vector<string> products;
        while (!canMake.empty()){
            string &target = recipes[canMake.back()];
            canMake.pop_back();
            products.push_back(target);
            for (int idx : leadsTo[target]){
                if (--indeg[idx] == 0){
                    canMake.push_back(idx);
                }
            }
        }
        return products;
    }
};
